File: optimization-tree-traversal-problems-complete/src/algorithms/treeTraversals.ts
```typescript
/**
 * src/algorithms/treeTraversals.ts
 *
 * This file contains implementations for various tree traversal algorithms (DFS and BFS based)
 * and common problems solved using these traversals.
 *
 * Problems Covered:
 * 1.  Standard DFS Traversals: Inorder, Preorder, Postorder (Recursive & Iterative)
 * 2.  Level Order Traversal (BFS)
 * 3.  Zigzag Level Order Traversal (BFS Variation)
 * 4.  Maximum Depth of Binary Tree (DFS)
 * 5.  Symmetric Tree Check (DFS)
 */

import { TreeNode } from '../data-structures/TreeNode';

// --- Problem 1: Standard DFS Traversals (Inorder, Preorder, Postorder) ---

/**
 * Performs an Inorder Traversal (Left -> Root -> Right) recursively.
 *
 * Inorder traversal for a BST visits nodes in ascending order.
 *
 * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is visited once.
 * Space Complexity: O(H) in the worst case (skewed tree) for the recursion stack, where H is the height of the tree.
 *                   O(log N) for a balanced tree.
 */
export function inorderTraversalRecursive<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    function traverse(node: TreeNode<T> | null) {
        if (node === null) {
            return;
        }
        traverse(node.left);
        result.push(node.val);
        traverse(node.right);
    }
    traverse(root);
    return result;
}

/**
 * Performs an Inorder Traversal (Left -> Root -> Right) iteratively using a stack.
 *
 * Algorithm:
 * 1. Initialize an empty stack and an empty result list.
 * 2. Start with the current node as the root.
 * 3. While the current node is not null or the stack is not empty:
 *    a. Keep pushing current node onto stack and move to its left child until current node is null.
 *    b. Pop a node from the stack. This is the next node in inorder sequence. Add its value to result.
 *    c. Move to the popped node's right child.
 *
 * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is pushed and popped once.
 * Space Complexity: O(H) in the worst case (skewed tree) for the stack, where H is the height of the tree.
 *                   O(log N) for a balanced tree.
 */
export function inorderTraversalIterative<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    const stack: TreeNode<T>[] = [];
    let current: TreeNode<T> | null = root;

    while (current !== null || stack.length > 0) {
        // Reach the leftmost node of the current subtree
        while (current !== null) {
            stack.push(current);
            current = current.left;
        }

        // Current must be null at this point, so we pop from stack
        current = stack.pop()!; // Guaranteed to not be null because of while condition
        result.push(current.val);

        // Move to the right subtree
        current = current.right;
    }
    return result;
}

/**
 * Performs a Preorder Traversal (Root -> Left -> Right) recursively.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(H) for recursion stack
 */
export function preorderTraversalRecursive<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    function traverse(node: TreeNode<T> | null) {
        if (node === null) {
            return;
        }
        result.push(node.val);
        traverse(node.left);
        traverse(node.right);
    }
    traverse(root);
    return result;
}

/**
 * Performs a Preorder Traversal (Root -> Left -> Right) iteratively using a stack.
 *
 * Algorithm:
 * 1. Initialize an empty stack and an empty result list.
 * 2. Push the root node onto the stack.
 * 3. While the stack is not empty:
 *    a. Pop a node from the stack.
 *    b. Add its value to the result list.
 *    c. Push its right child onto the stack (if exists).
 *    d. Push its left child onto the stack (if exists).
 *    (Push right first, then left, because stack is LIFO, so left will be processed before right).
 *
 * Time Complexity: O(N)
 * Space Complexity: O(H) for stack
 */
export function preorderTraversalIterative<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    if (root === null) {
        return result;
    }

    const stack: TreeNode<T>[] = [root];

    while (stack.length > 0) {
        const node = stack.pop()!;
        result.push(node.val);

        // Push right child first so left child is processed next (LIFO)
        if (node.right) {
            stack.push(node.right);
        }
        if (node.left) {
            stack.push(node.left);
        }
    }
    return result;
}

/**
 * Performs a Postorder Traversal (Left -> Right -> Root) recursively.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(H) for recursion stack
 */
export function postorderTraversalRecursive<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    function traverse(node: TreeNode<T> | null) {
        if (node === null) {
            return;
        }
        traverse(node.left);
        traverse(node.right);
        result.push(node.val);
    }
    traverse(root);
    return result;
}

/**
 * Performs a Postorder Traversal (Left -> Right -> Root) iteratively using two stacks.
 *
 * Algorithm:
 * 1. Initialize two stacks, `s1` and `s2`. Push root to `s1`.
 * 2. While `s1` is not empty:
 *    a. Pop a node from `s1` and push it onto `s2`.
 *    b. If the popped node has a left child, push it onto `s1`.
 *    c. If the popped node has a right child, push it onto `s1`.
 * 3. Once `s1` is empty, `s2` contains nodes in postorder (Root -> Right -> Left reversed).
 *    Pop all elements from `s2` and add their values to the result list.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(N) in worst case (skewed tree) for two stacks.
 */
export function postorderTraversalIterativeTwoStacks<T>(root: TreeNode<T> | null): T[] {
    const result: T[] = [];
    if (root === null) {
        return result;
    }

    const s1: TreeNode<T>[] = [root];
    const s2: TreeNode<T>[] = [];

    while (s1.length > 0) {
        const node = s1.pop()!;
        s2.push(node);

        if (node.left) {
            s1.push(node.left);
        }
        if (node.right) {
            s1.push(node.right);
        }
    }

    while (s2.length > 0) {
        result.push(s2.pop()!.val);
    }

    return result;
}

// --- Problem 2: Level Order Traversal (BFS) ---

/**
 * Performs a Level Order Traversal (Breadth-First Search) of the tree.
 * Returns a list of lists, where each inner list contains the nodes at that level.
 *
 * Algorithm:
 * 1. Initialize an empty queue and add the root node.
 * 2. While the queue is not empty:
 *    a. Get the number of nodes at the current level (queue size).
 *    b. Create a list for the current level's nodes.
 *    c. For each node at the current level:
 *       i. Dequeue the node.
 *       ii. Add its value to the current level list.
 *       iii. Enqueue its left child (if exists).
 *       iv. Enqueue its right child (if exists).
 *    d. Add the current level list to the overall result.
 *
 * Time Complexity: O(N), as each node is enqueued and dequeued exactly once.
 * Space Complexity: O(W) in the worst case, where W is the maximum width of the tree.
 *                   In a complete binary tree, W = N/2, so O(N).
 */
export function levelOrderTraversal<T>(root: TreeNode<T> | null): T[][] {
    const result: T[][] = [];
    if (root === null) {
        return result;
    }

    const queue: TreeNode<T>[] = [root];

    while (queue.length > 0) {
        const levelSize = queue.length;
        const currentLevel: T[] = [];

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()!;
            currentLevel.push(node.val);

            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }
        result.push(currentLevel);
    }
    return result;
}

// --- Problem 3: Zigzag Level Order Traversal ---

/**
 * Performs a Zigzag Level Order Traversal of the tree.
 * Returns a list of lists, where odd-numbered levels are traversed from left to right,
 * and even-numbered levels from right to left (0-indexed).
 *
 * Algorithm:
 * Uses a standard BFS queue, but alternates adding elements to the current level list.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(W) (maximum width of the tree)
 */
export function zigzagLevelOrderTraversal<T>(root: TreeNode<T> | null): T[][] {
    const result: T[][] = [];
    if (root === null) {
        return result;
    }

    const queue: TreeNode<T>[] = [root];
    let level = 0;

    while (queue.length > 0) {
        const levelSize = queue.length;
        const currentLevel: T[] = [];

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()!;
            // For even levels (0, 2, ...), add to end (left to right)
            // For odd levels (1, 3, ...), add to beginning (right to left)
            if (level % 2 === 0) {
                currentLevel.push(node.val);
            } else {
                currentLevel.unshift(node.val); // O(K) operation, where K is currentLevel size. Can be slow.
                                                 // Better: push to end, then reverse array at the end of loop.
            }

            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }
        result.push(currentLevel);
        level++;
    }
    return result;
}

// Optimized `zigzagLevelOrderTraversal` to avoid `unshift` performance penalty
export function zigzagLevelOrderTraversalOptimized<T>(root: TreeNode<T> | null): T[][] {
    const result: T[][] = [];
    if (root === null) {
        return result;
    }

    const queue: TreeNode<T>[] = [root];
    let level = 0;

    while (queue.length > 0) {
        const levelSize = queue.length;
        const currentLevel: T[] = [];

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()!;
            currentLevel.push(node.val); // Always add to end
            
            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }

        // If it's an odd level (1, 3, ...), reverse the collected list
        if (level % 2 === 1) {
            currentLevel.reverse();
        }
        result.push(currentLevel);
        level++;
    }
    return result;
}


// --- Problem 4: Maximum Depth of Binary Tree ---

/**
 * Calculates the maximum depth (height) of a binary tree recursively.
 * The maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
 *
 * Algorithm: Depth-First Search (DFS)
 * Base Case: If root is null, depth is 0.
 * Recursive Step: max_depth = 1 + max(max_depth(left_subtree), max_depth(right_subtree))
 *
 * Time Complexity: O(N), as each node is visited once.
 * Space Complexity: O(H) for the recursion stack in the worst case (skewed tree).
 *                   O(log N) for a balanced tree.
 */
export function maxDepthRecursive<T>(root: TreeNode<T> | null): number {
    if (root === null) {
        return 0;
    }
    const leftDepth = maxDepthRecursive(root.left);
    const rightDepth = maxDepthRecursive(root.right);
    return 1 + Math.max(leftDepth, rightDepth);
}

/**
 * Calculates the maximum depth (height) of a binary tree iteratively using BFS.
 *
 * Algorithm:
 * 1. Initialize a queue and add the root. If root is null, return 0.
 * 2. Initialize depth to 0.
 * 3. While the queue is not empty:
 *    a. Increment depth.
 *    b. Get the number of nodes at the current level (queue size).
 *    c. For each node at the current level:
 *       i. Dequeue the node.
 *       ii. Enqueue its left child (if exists).
 *       iii. Enqueue its right child (if exists).
 *
 * Time Complexity: O(N)
 * Space Complexity: O(W) (maximum width of the tree)
 */
export function maxDepthIterativeBFS<T>(root: TreeNode<T> | null): number {
    if (root === null) {
        return 0;
    }

    const queue: TreeNode<T>[] = [root];
    let depth = 0;

    while (queue.length > 0) {
        const levelSize = queue.length;
        depth++; // Increment depth for each level

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()!;
            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }
    }
    return depth;
}

// --- Problem 5: Symmetric Tree ---

/**
 * Checks if a binary tree is symmetric around its center.
 * This means the left and right subtrees must be mirror images of each other.
 *
 * Algorithm: Recursive DFS.
 * Define a helper function `isMirror(node1, node2)`:
 * 1. Base Case: If both nodes are null, they are symmetric (return true).
 * 2. Base Case: If one node is null and the other is not, they are not symmetric (return false).
 * 3. Recursive Step: Check if their values are equal AND
 *    `isMirror(node1.left, node2.right)` (outer children) AND
 *    `isMirror(node1.right, node2.left)` (inner children) are all true.
 *
 * Time Complexity: O(N), as each pair of nodes (one from left, one from right) is compared once.
 * Space Complexity: O(H) for the recursion stack.
 */
export function isSymmetricRecursive<T>(root: TreeNode<T> | null): boolean {
    if (root === null) {
        return true;
    }

    function isMirror(node1: TreeNode<T> | null, node2: TreeNode<T> | null): boolean {
        // Both are null, they are symmetric
        if (node1 === null && node2 === null) {
            return true;
        }
        // One is null, the other is not, not symmetric
        if (node1 === null || node2 === null) {
            return false;
        }
        // Check values and mirror children
        return (
            node1.val === node2.val &&
            isMirror(node1.left, node2.right) && // Outer children
            isMirror(node1.right, node2.left)    // Inner children
        );
    }

    return isMirror(root.left, root.right);
}

/**
 * Checks if a binary tree is symmetric iteratively using a queue (BFS-like approach).
 *
 * Algorithm:
 * 1. If root is null, return true.
 * 2. Initialize a queue and add `root.left` and `root.right`.
 * 3. While the queue is not empty:
 *    a. Dequeue two nodes, `t1` and `t2`.
 *    b. If both are null, continue (they are symmetric).
 *    c. If one is null and the other is not, or their values are different, return false.
 *    d. Enqueue `t1.left`, `t2.right` (outer children).
 *    e. Enqueue `t1.right`, `t2.left` (inner children).
 * 4. If the loop completes, the tree is symmetric, return true.
 *
 * Time Complexity: O(N)
 * Space Complexity: O(W) (maximum width of the tree)
 */
export function isSymmetricIterative<T>(root: TreeNode<T> | null): boolean {
    if (root === null) {
        return true;
    }

    const queue: (TreeNode<T> | null)[] = [];
    queue.push(root.left);
    queue.push(root.right);

    while (queue.length > 0) {
        const t1 = queue.shift();
        const t2 = queue.shift();

        // Both are null, continue checking other pairs
        if (t1 === null && t2 === null) {
            continue;
        }
        // One is null or values are different, not symmetric
        if (t1 === null || t2 === null || t1.val !== t2.val) {
            return false;
        }

        // Enqueue outer children pair
        queue.push(t1.left);
        queue.push(t2.right);

        // Enqueue inner children pair
        queue.push(t1.right);
        queue.push(t2.left);
    }

    return true;
}
```

File: algorithm-tree-traversal-implementation-complete/src/main/java/tree_traversals/TreeTraversals.java
```java
package tree_traversals;

import java.util.*;

/**
 * This class provides implementations for various binary tree traversal problems.
 * It includes classic DFS (Inorder, Preorder, Postorder) using both recursive and iterative approaches,
 * and BFS (Level Order, Zigzag, Right Side View, Vertical Order) traversals.
 */
public class TreeTraversals {

    // --- Problem 1: Standard Traversals (Inorder, Preorder, Postorder) ---

    // --- Inorder Traversal (Left -> Root -> Right) ---

    /**
     * Performs an Inorder Traversal recursively.
     * The order of visiting nodes is Left child, then Root, then Right child.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the inorder traversal of the tree.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is visited once.
     * Space Complexity: O(H), where H is the height of the tree. This is due to the recursion stack.
     *                   In the worst case (skewed tree), H can be N, so O(N).
     *                   In the best case (balanced tree), H is log N, so O(log N).
     */
    public List<Integer> inorderTraversalRecursive(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        inorderHelperRecursive(root, result);
        return result;
    }

    private void inorderHelperRecursive(TreeNode node, List<Integer> result) {
        if (node == null) {
            return;
        }
        inorderHelperRecursive(node.left, result);  // Traverse left subtree
        result.add(node.val);                       // Visit root
        inorderHelperRecursive(node.right, result); // Traverse right subtree
    }

    /**
     * Performs an Inorder Traversal iteratively using a stack.
     * This approach mimics the recursion call stack explicitly.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the inorder traversal of the tree.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is pushed and popped once.
     * Space Complexity: O(H), where H is the height of the tree. This is due to the stack.
     *                   In the worst case (skewed tree), H can be N, so O(N).
     *                   In the best case (balanced tree), H is log N, so O(log N).
     */
    public List<Integer> inorderTraversalIterative(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode current = root;

        while (current != null || !stack.isEmpty()) {
            // Traverse to the leftmost node, pushing all visited nodes onto the stack
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            // Pop the top node (which is the leftmost unvisited node)
            current = stack.pop();
            result.add(current.val); // Visit the node

            // Now, process its right subtree
            current = current.right;
        }
        return result;
    }

    // --- Preorder Traversal (Root -> Left -> Right) ---

    /**
     * Performs a Preorder Traversal recursively.
     * The order of visiting nodes is Root, then Left child, then Right child.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the preorder traversal of the tree.
     *
     * Time Complexity: O(N). Each node is visited once.
     * Space Complexity: O(H) for recursion stack.
     */
    public List<Integer> preorderTraversalRecursive(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        preorderHelperRecursive(root, result);
        return result;
    }

    private void preorderHelperRecursive(TreeNode node, List<Integer> result) {
        if (node == null) {
            return;
        }
        result.add(node.val);                       // Visit root
        preorderHelperRecursive(node.left, result);  // Traverse left subtree
        preorderHelperRecursive(node.right, result); // Traverse right subtree
    }

    /**
     * Performs a Preorder Traversal iteratively using a stack.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the preorder traversal of the tree.
     *
     * Time Complexity: O(N). Each node is pushed and popped once.
     * Space Complexity: O(H) for stack.
     */
    public List<Integer> preorderTraversalIterative(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            TreeNode current = stack.pop();
            result.add(current.val); // Visit the node

            // Push right child first, then left child.
            // This ensures left child is processed before right child
            // because stack is LIFO.
            if (current.right != null) {
                stack.push(current.right);
            }
            if (current.left != null) {
                stack.push(current.left);
            }
        }
        return result;
    }

    // --- Postorder Traversal (Left -> Right -> Root) ---

    /**
     * Performs a Postorder Traversal recursively.
     * The order of visiting nodes is Left child, then Right child, then Root.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the postorder traversal of the tree.
     *
     * Time Complexity: O(N). Each node is visited once.
     * Space Complexity: O(H) for recursion stack.
     */
    public List<Integer> postorderTraversalRecursive(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorderHelperRecursive(root, result);
        return result;
    }

    private void postorderHelperRecursive(TreeNode node, List<Integer> result) {
        if (node == null) {
            return;
        }
        postorderHelperRecursive(node.left, result);  // Traverse left subtree
        postorderHelperRecursive(node.right, result); // Traverse right subtree
        result.add(node.val);                       // Visit root
    }

    /**
     * Performs a Postorder Traversal iteratively using two stacks.
     * This is a common method for iterative postorder.
     * The first stack is used for standard preorder-like processing (Root, Right, Left).
     * The second stack collects the elements in reverse postorder (Root, Right, Left becomes Left, Right, Root when popped).
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the postorder traversal of the tree.
     *
     * Time Complexity: O(N). Each node is pushed and popped twice (once from s1, once from s2).
     * Space Complexity: O(H) for both stacks.
     */
    public List<Integer> postorderTraversalIterativeTwoStacks(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Stack<TreeNode> s1 = new Stack<>(); // Main stack for processing
        Stack<TreeNode> s2 = new Stack<>(); // Stack to store result in reverse order

        s1.push(root);

        while (!s1.isEmpty()) {
            TreeNode current = s1.pop();
            s2.push(current); // Push to s2 (will be popped in L->R->Root order)

            // Push left child first, then right child to s1.
            // This ensures right child is processed before left child on s1's next pop.
            if (current.left != null) {
                s1.push(current.left);
            }
            if (current.right != null) {
                s1.push(current.right);
            }
        }

        // Pop all elements from s2 to get the final postorder traversal
        while (!s2.isEmpty()) {
            result.add(s2.pop().val);
        }
        return result;
    }

    /**
     * Performs a Postorder Traversal iteratively using one stack.
     * This method is more complex but uses less space than two stacks.
     * It relies on tracking the `lastVisited` node to determine if the right child has been processed.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the postorder traversal of the tree.
     *
     * Time Complexity: O(N). Each node is visited multiple times (pushed, popped, peeked).
     * Space Complexity: O(H) for the stack.
     */
    public List<Integer> postorderTraversalIterativeOneStack(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Stack<TreeNode> stack = new Stack<>();
        TreeNode current = root;
        TreeNode lastVisited = null; // Tracks the last node added to the result

        while (current != null || !stack.isEmpty()) {
            // Traverse to the leftmost node
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            // Peek at the top of the stack
            TreeNode peekNode = stack.peek();

            // If the right child exists and hasn't been visited yet, go right
            if (peekNode.right != null && peekNode.right != lastVisited) {
                current = peekNode.right;
            } else {
                // Otherwise, pop the node (both left and right subtrees processed or null)
                result.add(peekNode.val);
                lastVisited = stack.pop(); // Mark as last visited
            }
        }
        return result;
    }

    // --- Problem 2: Level Order Traversal (BFS) ---

    /**
     * Performs a Level Order Traversal (Breadth-First Search).
     * Visits nodes level by level from left to right.
     *
     * @param root The root of the binary tree.
     * @return A list of lists of integers, where each inner list represents a level of the tree.
     *
     * Time Complexity: O(N). Each node is visited and enqueued/dequeued exactly once.
     * Space Complexity: O(W), where W is the maximum width of the tree. In the worst case (complete binary tree),
     *                   W can be N/2, so O(N). In the best case (skewed tree), W is 1, so O(1).
     */
    public List<List<Integer>> levelOrderTraversal(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size(); // Number of nodes at the current level
            List<Integer> currentLevelNodes = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode current = queue.poll();
                currentLevelNodes.add(current.val);

                if (current.left != null) {
                    queue.offer(current.left);
                }
                if (current.right != null) {
                    queue.offer(current.right);
                }
            }
            result.add(currentLevelNodes);
        }
        return result;
    }

    // --- Problem 3: Zigzag Level Order Traversal ---

    /**
     * Performs a Zigzag Level Order Traversal.
     * Nodes are traversed level by level, but the order alternates:
     * - Level 0: Left to Right
     * - Level 1: Right to Left
     * - Level 2: Left to Right
     * ... and so on.
     *
     * This is achieved using a Deque (double-ended queue) for each level,
     * or by adding elements to the result list from the beginning/end based on level parity.
     *
     * @param root The root of the binary tree.
     * @return A list of lists of integers, representing the zigzag traversal.
     *
     * Time Complexity: O(N). Each node is visited and enqueued/dequeued once.
     * Space Complexity: O(W), where W is the maximum width of the tree (for the queue).
     */
    public List<List<Integer>> zigzagLevelOrderTraversal(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean leftToRight = true; // Direction flag for current level

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            // Use a LinkedList to efficiently add elements to front/back
            LinkedList<Integer> currentLevelNodes = new LinkedList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode current = queue.poll();

                if (leftToRight) {
                    currentLevelNodes.addLast(current.val); // Add to end for L-R
                } else {
                    currentLevelNodes.addFirst(current.val); // Add to front for R-L
                }

                if (current.left != null) {
                    queue.offer(current.left);
                }
                if (current.right != null) {
                    queue.offer(current.right);
                }
            }
            result.add(currentLevelNodes);
            leftToRight = !leftToRight; // Toggle direction for next level
        }
        return result;
    }

    // --- Problem 4: Binary Tree Right Side View ---

    /**
     * Given the `root` of a binary tree, imagine you stand on the right side of it,
     * return the values of the nodes you can see ordered from top to bottom.
     * This can be solved efficiently using a Level Order Traversal (BFS).
     * For each level, the last node encountered (or the first node if traversing right-to-left)
     * will be the one visible from the right side.
     *
     * @param root The root of the binary tree.
     * @return A list of integers representing the right side view of the tree.
     *
     * Time Complexity: O(N). Each node is visited and enqueued/dequeued once.
     * Space Complexity: O(W) for the queue.
     */
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            // The value of the first node popped from the queue at each level (when processing left to right)
            // will be the rightmost node for that level, because we add children left then right.
            // If we add nodes to queue from left to right, then the last node processed for a level will be visible.
            // A simpler approach: add the value of the first node of each level to the result list.
            // Wait, this is `right side view`. So, for each level, we want the rightmost node.
            // If we traverse L->R for a level, the last element processed is the rightmost.
            // Or, more robust: iterate from right to left in the current level.
            // Let's stick with the common pattern: iterate L->R, and the LAST node processed is the rightmost.

            for (int i = 0; i < levelSize; i++) {
                TreeNode current = queue.poll();

                // If it's the last node of the current level, it's visible from the right side
                if (i == levelSize - 1) {
                    result.add(current.val);
                }

                // Enqueue children for the next level (left first, then right)
                if (current.left != null) {
                    queue.offer(current.left);
                }
                if (current.right != null) {
                    queue.offer(current.right);
                }
            }
        }
        return result;
    }


    // --- Problem 5: Vertical Order Traversal ---

    /**
     * Performs a Vertical Order Traversal of a binary tree.
     * Nodes are ordered by their column index. Within each column, nodes are ordered by row index (depth).
     * If nodes are at the same column and row, they are ordered by their value.
     *
     * This problem requires tracking both the node itself and its column index (horizontal distance from root).
     * A BFS approach is suitable, using a Queue to store `(Node, Column)` pairs.
     * A `TreeMap` is used to store nodes grouped by column, ensuring columns are naturally sorted.
     * Each column then stores a list of nodes, which can be further sorted by row then value.
     *
     * @param root The root of the binary tree.
     * @return A list of lists of integers, representing the vertical traversal.
     *
     * Time Complexity: O(N log W) where N is the number of nodes and W is the width of the tree.
     *                  N nodes are processed. For each node, map operations take log W time (if using TreeMap),
     *                  and sorting within columns can add to this if there are many nodes at the same column.
     *                  If a simple HashMap is used and then keys are sorted: O(N + W log W + N log D) where D is max nodes in a column.
     * Space Complexity: O(N) for the map and queue.
     */
    public List<List<Integer>> verticalOrderTraversal(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        // TreeMap to store nodes by column index.
        // Key: Column index (horizontal distance from root, root is 0).
        // Value: Another TreeMap to store nodes by row index (depth).
        // Key: Row index (depth from root, root is 0).
        // Value: List of node values at that specific column and row.
        TreeMap<Integer, TreeMap<Integer, List<Integer>>> columns = new TreeMap<>();
        Queue<NodeInfo> queue = new LinkedList<>();

        // Start with the root at column 0, row 0
        queue.offer(new NodeInfo(root, 0, 0));

        // Keep track of min and max column indices to iterate through later
        int minCol = 0, maxCol = 0;

        while (!queue.isEmpty()) {
            NodeInfo current = queue.poll();
            TreeNode node = current.node;
            int col = current.col;
            int row = current.row;

            // Update min/max column indices
            minCol = Math.min(minCol, col);
            maxCol = Math.max(maxCol, col);

            // Add node to the map
            columns.putIfAbsent(col, new TreeMap<>());
            columns.get(col).putIfAbsent(row, new ArrayList<>());
            columns.get(col).get(row).add(node.val);

            if (node.left != null) {
                queue.offer(new NodeInfo(node.left, col - 1, row + 1));
            }
            if (node.right != null) {
                queue.offer(new NodeInfo(node.right, col + 1, row + 1));
            }
        }

        // Construct the final result list
        for (int col = minCol; col <= maxCol; col++) {
            List<Integer> verticalList = new ArrayList<>();
            // Get all nodes for the current column, ordered by row
            if (columns.containsKey(col)) {
                for (List<Integer> nodesAtRow : columns.get(col).values()) {
                    // For nodes at the same column and row, sort by value as per problem statement
                    Collections.sort(nodesAtRow);
                    verticalList.addAll(nodesAtRow);
                }
            }
            result.add(verticalList);
        }

        return result;
    }

    // Helper class to store Node with its column and row indices
    private static class NodeInfo {
        TreeNode node;
        int col;
        int row;

        NodeInfo(TreeNode node, int col, int row) {
            this.node = node;
            this.col = col;
            this.row = row;
        }
    }
}
```